Types: allow for arbitrary theme values (for 3rd party plugins)

RobinMalfait opened the pull request:
This PR provides more improvements to the types config.

- Allows for arbitrary objects in the `theme` section for third party plugins.
- ~~Provides code completion for the `theme()` helper function~~

<!--

👋 Hey, thanks for your interest in contributing to Tailwind!

**Please ask first before starting work on any significant new features.**

It's never a fun experience to have your pull request declined after investing a lot of time and effort into a new feature. To avoid this from happening, we request that contributors create an issue to first discuss any significant new features. This includes things like adding new utilities, creating new at-rules, or adding new component examples to the documentation.

https://github.com/tailwindcss/tailwindcss/blob/master/.github/CONTRIBUTING.md

-->

RobinMalfait committed:
allow for arbitrary configuration in the `theme` section

This is useful for third party plugins otherwise you will get an error.

RobinMalfait committed:
WIP: `theme()` utility function code completion

This will give you code completion in the `theme()` function. The reason
it is still a WIP is that this only works with the default config right
now and not 100% sure if it is possible to define generics in JSDoc.

The idea would be to:
- Provide types from the default config
- Provide types from the custom config (e.g.: 3rd party plugin)
- Override default config types with the overrides of the user's config

Right now this only provides types for the defaultConfig which might
result in dropping all of this in favor of a much simpler:

```ts
theme<D = any>(path: string, defaultValue: D) => D
```

But this sadly doesn't give you "nice" auto completion. However, the
default might be good enough if we don't error on for example
`theme('customPlugin')` which is currently not the case.

RobinMalfait committed:
update changelog

anosadal commented:
Looks like your django.root has a trailing slash on it, which is mentioned in the documentation as not the way to go.
If that's not the case, please print out what req.uri is in that case. I can't immediately see why that should be happening and your fix isn't the right approach (it's symptom patching, not fixing the problem).

RobinMalfait commented:
@anosadal everything alright buddy?

RobinMalfait committed:
undo all `theme` types, and type it as `theme(path: string): any`

Since currently we don't want to investigate time to make the code
completion *perfect* (because it might be even impossible to do it
properly due to resolving of overrides, extend and deeply nested presets)

For now we will provide a way simpler type, which is better than
incorrect types. So far we only had types for the default config theme
*only*.
